pkg/bbr/framework/registry.go

```diff
@@ -0,0 +1,276 @@
+	if _, ok := r.pluginsFactory[plugin.TypedName().Type]; !ok {
+		err := fmt.Errorf("no plugin factory registered for plugin interface type %s", plugin.TypedName().Type)
+		return err
+	}
+	r.plugins[plugin.TypedName().Type] = plugin
+
+	return nil
+}
+
+// Retrieves a plugin factory by type key
+func (r *pluginRegistry) GetFactory(typeKey string) (PluginFactoryFunc, error) {
+	if pluginFactory, ok := r.pluginsFactory[typeKey]; ok {
+		return pluginFactory, nil
+	}
+	return nil, fmt.Errorf("plugin type %s not found", typeKey)
+}
+
+// Retrieves a plugin instance by type key
+func (r *pluginRegistry) GetPlugin(typeKey string) (bbrplugins.BBRPlugin, error) {
+	if plugin, ok := r.plugins[typeKey]; ok {
+		return plugin, nil
+	}
+	return nil, fmt.Errorf("plugin type %s not found", typeKey)
+}
+
+// Constructs a new plugin (a caller can perform either type  assertion of a concrete implementation of the BBR plugin)
+func (r *pluginRegistry) CreatePlugin(typeKey string) (bbrplugins.BBRPlugin, error) {
+	if factory, ok := r.pluginsFactory[typeKey]; ok {
+		plugin := factory()
+		return plugin, nil
+	}
+	return nil, fmt.Errorf("plugin %s not registered", typeKey)
+}
+
+// Removes a plugin factory by type key
+func (r *pluginRegistry) UnregisterFactory(typeKey string) error {
+	if _, ok := r.pluginsFactory[typeKey]; ok {
+		delete(r.pluginsFactory, typeKey)
+		return nil
+	}
+	return fmt.Errorf("plugin (%s) not found", typeKey)
+}
+
+// ListPlugins lists all registered plugins
+func (r *pluginRegistry) ListPlugins() []string {
+	typeKeys := make([]string, 0, len(r.plugins))
+	for k := range r.plugins {
+		typeKeys = append(typeKeys, k)
+	}
+	return typeKeys
+}
+
+// ListPlugins lists all registered plugins; this functionis not really needed. Just for sanity checks and tests
+func (r *pluginRegistry) ListFactories() []string {
+	typeKeys := make([]string, 0, len(r.pluginsFactory))
+	for k := range r.pluginsFactory {
+		typeKeys = append(typeKeys, k)
+	}
+	return typeKeys
+}
+
+// Get factories
+func (r *pluginRegistry) GetFactories() map[string]PluginFactoryFunc {
+	return r.pluginsFactory
+}
+
+// Get plugins
+func (r *pluginRegistry) GetPlugins() map[string]bbrplugins.BBRPlugin {
+	return r.plugins
+}
+
+// Checks for presense of a factory in this registry
+func (r *pluginRegistry) ContainsFactory(typeKey string) bool {
+	_, exists := r.pluginsFactory[typeKey]
+	return exists
+}
+
+// Helper: Checks for presense of a plugin in this registry
+func (r *pluginRegistry) ContainsPlugin(typeKey string) bool {
+	_, exists := r.plugins[typeKey]
+	return exists
+}
+
+func (r *pluginRegistry) String() string {
+	return fmt.Sprintf("{plugins=%v}{pluginsFactory=%v}", r.plugins, r.pluginsFactory)
+}
+
+//-------------------------- PluginsChain implementation --------------------------
+
+// PluginsChain is a sequence of plugins to be executed in order inside the ext_proc server
```

Comment:
The code currently has a single request chain only, consider removing types and variables that suggest multiplicity in order to make this PR smaller and more focused.